### ISSUE TITLE
Resolver difficulties

### ISSUE THREAD

**Issue by mbauman** [JuliaLang/Pkg.jl#303]
I'm constantly running into difficulties with the resolver.  It all seems to stem from upper-bounding Compat for StatsBase.

```
$ mv ~/.julia ~/.julia-broken # remove .julia to start from a blank slate

[master][arctic1:~/julia]
$ ./julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0-DEV.5219 (2018-05-25 14:37 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit b65712285a (0 days old master)
|__/                   |  x86_64-linux-gnu

(v0.7) pkg> add BenchmarkTools
┌ Info: Pkg is running without precompile statements, first action will be slow.
│ Rebuild julia with the environment variable `JULIA_PKG3_PRECOMPILE` set to enable precompilation of Pkg.
└ This message can be disabled by setting the env variable `JULIA_PKG3_DISABLE_PRECOMPILE_WARNING`.
   Cloning default registries into /home/mbauman/.julia/registries
   Cloning registry Uncurated from "https://github.com/JuliaRegistries/Uncurated.git"
  Updating registry at `~/.julia/registries/Uncurated`
  Updating git-repo `https://github.com/JuliaRegistries/Uncurated.git`
 Resolving package versions...
ERROR: Resolve failed to satisfy requirements for package BinDeps [9e28174c]:
 BinDeps [9e28174c] log:
 ├─possible versions are: [0.0.1, 0.2.0-0.2.14, 0.3.0-0.3.23, 0.4.0-0.4.5, 0.4.7, 0.5.0, 0.6.0, 0.7.0, 0.8.0-0.8.8] or uninstalled
 ├─restricted by julia compatibility requirements to versions: [0.2.6, 0.2.14, 0.3.7, 0.3.15, 0.3.23, 0.5.0, 0.8.5-0.8.8] or uninstalled
 ├─versions reduced by equivalence to: [0.2.14, 0.3.7, 0.3.23, 0.5.0, 0.8.5, 0.8.7-0.8.8] or uninstalled
 ├─restricted by compatibility requirements with (unknown) [8f399da3] to versions: uninstalled
 │ └─(unknown) [8f399da3] log:
 │   ├─possible versions are: 0.7.0 or uninstalled
 │   └─fixed by the MaxSum heuristic to version 0.7.0-DEV.5219
 └─found to have no compatible versions left with HDF5 [f67ccb44]
   └─HDF5 [f67ccb44] log:
     ├─possible versions are: [0.1.0, 0.2.0-0.2.31, 0.3.0, 0.4.0-0.4.18, 0.5.0-0.5.8, 0.6.0-0.6.6, 0.7.0-0.7.3, 0.8.0-0.8.8, 0.9.1-0.9.2] or uninstalled
     ├─restricted by julia compatibility requirements to versions: [0.2.6, 0.2.14, 0.2.22, 0.2.30, 0.4.5, 0.4.13, 0.5.2, 0.6.1, 0.7.2, 0.8.6-0.8.8, 0.9.1-0.9.2] or uninstalled
     ├─restricted by compatibility requirements with JLD [4138dd39] to versions: [0.2.6, 0.2.14, 0.2.22, 0.2.30, 0.4.5, 0.4.13, 0.5.2, 0.6.1, 0.7.2, 0.8.6-0.8.8, 0.9.1-0.9.2]
     │ └─JLD [4138dd39] log:
     │   ├─possible versions are: [0.5.0-0.5.9, 0.6.0-0.6.11, 0.8.0-0.8.3] or uninstalled
     │   ├─restricted by compatibility requirements with BenchmarkTools [6e4b80f9] to versions: [0.5.7, 0.6.5, 0.8.1-0.8.3] or uninstalled
     │   │ └─BenchmarkTools [6e4b80f9] log:
     │   │   ├─possible versions are: [0.0.1-0.0.8, 0.2.0-0.2.5, 0.3.0] or uninstalled
     │   │   ├─restricted to versions * by an explicit requirement, leaving only versions [0.0.1-0.0.8, 0.2.0-0.2.5, 0.3.0]
     │   │   ├─restricted by julia compatibility requirements to versions: 0.0.8 or uninstalled, leaving only versions: 0.0.8
     │   │   └─fixed during graph pruning to its only remaining available version, 0.0.8
     │   ├─restricted by julia compatibility requirements to versions: [0.6.5, 0.8.1-0.8.3] or uninstalled
     │   ├─restricted by compatibility requirements with BenchmarkTools [6e4b80f9] to versions: 0.8.1-0.8.3
     │   │ └─BenchmarkTools [6e4b80f9] log: see above
     │   └─versions reduced by equivalence to: [0.8.1, 0.8.3]
     ├─versions reduced by equivalence to: [0.2.6, 0.2.14, 0.2.30, 0.4.5, 0.6.1, 0.7.2, 0.8.8, 0.9.2]
     └─restricted by compatibility requirements with (unknown) [8f399da3] to versions: 0.9.2 or uninstalled, leaving only versions: 0.9.2
       └─(unknown) [8f399da3] log: see above

Stacktrace:
 [1] #propagate_constraints!#61(::Bool, ::Function, ::Pkg.GraphType.Graph, ::Set{Int64}) at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/GraphType.jl:1005
 [2] propagate_constraints! at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/GraphType.jl:946 [inlined]
 [3] #simplify_graph!#121(::Bool, ::Function, ::Pkg.GraphType.Graph, ::Set{Int64}) at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/GraphType.jl:1460
 [4] simplify_graph! at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/GraphType.jl:1460 [inlined]
 [5] trigger_failure!(::Pkg.GraphType.Graph, ::Array{Int64,1}, ::Tuple{Int64,Int64}) at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/Resolve.jl:407
 [6] resolve(::Pkg.GraphType.Graph) at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/Resolve.jl:36
 [7] resolve_versions!(::Pkg.Types.Context, ::Array{Pkg.Types.PackageSpec,1}) at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/Operations.jl:288
 [8] #add_or_develop#50(::Array{Base.UUID,1}, ::Function, ::Pkg.Types.Context, ::Array{Pkg.Types.PackageSpec,1}) at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/Operations.jl:965
 [9] #add_or_develop at ./<missing>:0 [inlined]
 [10] #add_or_develop#8(::Symbol, ::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Pkg.Types.Context, ::Array{Pkg.Types.PackageSpec,1}) at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/API.jl:38
 [11] #add_or_develop at ./<missing>:0 [inlined]
 [12] do_add_or_develop!(::Pkg.Types.Context, ::Array{Union{Pkg.Types.VersionRange, String, Pkg.REPLMode.Command, Pkg.REPLMode.Option, Pkg.REPLMode.Rev},1}, ::Pkg.REPLMode.CommandKind) at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/REPLMode.jl:594
 [13] #invokelatest#1 at ./essentials.jl:667 [inlined]
 [14] invokelatest at ./essentials.jl:666 [inlined]
 [15] do_cmd!(::Array{Union{Pkg.Types.VersionRange, String, Pkg.REPLMode.Command, Pkg.REPLMode.Option, Pkg.REPLMode.Rev},1}, ::REPL.LineEditREPL) at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/REPLMode.jl:274
 [16] #do_cmd#8(::Bool, ::Function, ::REPL.LineEditREPL, ::String) at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/REPLMode.jl:234
 [17] do_cmd at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/REPLMode.jl:231 [inlined]
 [18] (::getfield(Pkg.REPLMode, Symbol("##27#30")){REPL.LineEditREPL,REPL.LineEdit.Prompt})(::REPL.LineEdit.MIState, ::Base.GenericIOBuffer{Array{UInt8,1}}, ::Bool) at /home/mbauman/julia/usr/share/julia/stdlib/v0.7/Pkg/src/REPLMode.jl:949
 [19] top-level scope
```

**Comment by StefanKarpinski** [comment]
I suspect that we're hitting a fundamental problem with the current resolver. Upper bounds are what make version resolution NP-hard and we've never really had very many of them in the past, which means that we were never really testing the resolver against hard problems. It's possible that we may be able to fix this without a complete rewrite but I'm a bit skeptical. My guess is that we need a SAT solver.

**Comment by KristofferC** [comment]
I get:

```
shell> mkdir NewEnv
mkdir: NewEnv: File exists

shell> cd NewEnv
/Users/kristoffer/julia/NewEnv

(julia) pkg> init
Initialized project at /Users/kristoffer/julia/NewEnv/Project.toml

(NewEnv) pkg> add BenchmarkTools
  Updating registry at `~/.julia/registries/Uncurated`
  Updating git-repo `https://github.com/JuliaRegistries/Uncurated.git`
 Resolving package versions...
  Updating `Project.toml`
  [6e4b80f9] + BenchmarkTools v0.3.0
  Updating `Manifest.toml`
  [6e4b80f9] + BenchmarkTools v0.3.0
  [34da2185] + Compat v0.65.1
  [682c06a0] + JSON v0.17.2
  [4d1e1d77] + Nullables v0.0.5
...
```

**Comment by mbauman** [comment]
```julia
shell> mkdir NewEnv

shell> cd NewEnv
/home/mbauman/julia/NewEnv

(v0.7) pkg> init
┌ Info: Pkg is running without precompile statements, first action will be slow.
│ Rebuild julia with the environment variable `JULIA_PKG3_PRECOMPILE` set to enable precompilation of Pkg.
└ This message can be disabled by setting the env variable `JULIA_PKG3_DISABLE_PRECOMPILE_WARNING`.
Initialized project at /home/mbauman/julia/NewEnv/Project.toml

(NewEnv) pkg> add BenchmarkTools
  Updating registry at `~/.julia/registries/Uncurated`
  Updating git-repo `https://github.com/JuliaRegistries/Uncurated.git`
 Resolving package versions...
    JULIA usr/lib/julia/sys-o.a
ERROR: Resolve failed to satisfy requirements for package BinDeps [9e28174c]:
 BinDeps [9e28174c] log:
 ├─possible versions are: [0.0.1, 0.2.0-0.2.14, 0.3.0-0.3.23, 0.4.0-0.4.5, 0.4.7, 0.5.0, 0.6.0, 0.7.0, 0.8.0-0.8.8] or uninstalled
 ├─restricted by julia compatibility requirements to versions: [0.2.6, 0.2.14, 0.3.7, 0.3.15, 0.3.23, 0.5.0, 0.8.5-0.8.8] or uninstalled
 ├─versions reduced by equivalence to: [0.2.14, 0.3.7, 0.3.23, 0.5.0, 0.8.5, 0.8.7-0.8.8] or uninstalled
 ├─restricted by compatibility requirements with (unknown) [8f399da3] to versions: uninstalled
 │ └─(unknown) [8f399da3] log:
 │   ├─possible versions are: 0.7.0 or uninstalled
 │   └─fixed by the MaxSum heuristic to version 0.7.0-DEV.5224
 └─found to have no compatible versions left with HDF5 [f67ccb44]
   └─HDF5 [f67ccb44] log:
     ├─possible versions are: [0.1.0, 0.2.0-0.2.31, 0.3.0, 0.4.0-0.4.18, 0.5.0-0.5.8, 0.6.0-0.6.6, 0.7.0-0.7.3, 0.8.0-0.8.8, 0.9.1-0.9.2] or uninstalled
     ├─restricted by julia compatibility requirements to versions: [0.2.6, 0.2.14, 0.2.22, 0.2.30, 0.4.5, 0.4.13, 0.5.2, 0.6.1, 0.7.2, 0.8.6-0.8.8, 0.9.1-0.9.2] or uninstalled
     ├─restricted by compatibility requirements with JLD [4138dd39] to versions: [0.2.6, 0.2.14, 0.2.22, 0.2.30, 0.4.5, 0.4.13, 0.5.2, 0.6.1, 0.7.2, 0.8.6-0.8.8, 0.9.1-0.9.2]
     │ └─JLD [4138dd39] log:
     │   ├─possible versions are: [0.5.0-0.5.9, 0.6.0-0.6.11, 0.8.0-0.8.3] or uninstalled
     │   ├─restricted by compatibility requirements with BenchmarkTools [6e4b80f9] to versions: [0.5.7, 0.6.5, 0.8.1-0.8.3] or uninstalled
     │   │ └─BenchmarkTools [6e4b80f9] log:
     │   │   ├─possible versions are: [0.0.1-0.0.8, 0.2.0-0.2.5, 0.3.0] or uninstalled
     │   │   ├─restricted to versions * by an explicit requirement, leaving only versions [0.0.1-0.0.8, 0.2.0-0.2.5, 0.3.0]
     │   │   ├─restricted by julia compatibility requirements to versions: 0.0.8 or uninstalled, leaving only versions: 0.0.8
     │   │   └─fixed during graph pruning to its only remaining available version, 0.0.8
     │   ├─restricted by julia compatibility requirements to versions: [0.6.5, 0.8.1-0.8.3] or uninstalled
     │   ├─restricted by compatibility requirements with BenchmarkTools [6e4b80f9] to versions: 0.8.1-0.8.3
     │   │ └─BenchmarkTools [6e4b80f9] log: see above
     │   └─versions reduced by equivalence to: [0.8.1, 0.8.3]
     ├─versions reduced by equivalence to: [0.2.6, 0.2.14, 0.2.30, 0.4.5, 0.6.1, 0.7.2, 0.8.8, 0.9.2]
     └─restricted by compatibility requirements with (unknown) [8f399da3] to versions: 0.9.2 or uninstalled, leaving only versions: 0.9.2
       └─(unknown) [8f399da3] log: see above

Stacktrace:
...
```

Doesn't the resolver use some randomness in its initial state?

**Comment by KristofferC** [comment]
I ran

```
for i in 1:100
    Pkg3.rm("BenchmarkTools")
    Pkg3.add("BenchmarkTools")
end
```

and it never failed. Spooky...

**Comment by KristofferC** [comment]
I am confused where `(unknown) [8f399da3]` comes from. There is a bug when adding unregistered packages that fails to show the name but here we should only have registered packages...

**Comment by mbauman** [comment]
I was just looking at that, too.  `ack 8f399da3` shows that it's coming from `Libdl`.

**Comment by KristofferC** [comment]
Perhaps we are not detecting it as a stdlib...

What if you do `Pkg.Context().stdlibs`, is it there?

**Comment by mbauman** [comment]
```
julia> Pkg.Context().stdlibs
WARNING: Base.Pkg is deprecated, run `using Pkg` instead
 in module Main
Dict{Base.UUID,String} with 30 entries:
  UUID("37e2e46d-f89d-539d-b4ee-838fcccc9c8e") => "LinearAlgebra"
  UUID("2f01184e-e22b-5df5-ae63-d93ebab69eaf") => "SparseArrays"
  UUID("7b1f6079-737a-58dc-b8bc-7a2ca5c1b5ee") => "FileWatching"
  UUID("8f399da3-3557-5675-b5ff-fb832c97cbdb") => "Libdl"
...
```

**Comment by KristofferC** [comment]
Libdl and BinDeps reminds me of:

https://github.com/JuliaLang/Pkg.jl/pull/166

and

https://github.com/JuliaLang/Pkg.jl/blob/d90fdeee8dc9a44c2850bcffe320a2516519b4a6/bin/generate.jl#L90-L93

**Comment by mbauman** [comment]
Forgive my ignorance here, but is this an issue:

```
 │   ├─possible versions are: 0.7.0 or uninstalled
 │   └─fixed by the MaxSum heuristic to version 0.7.0-DEV.5220
```

I've seen this in other resolver failures (although typically with a package on a master branch and not julia itself, listing just `vX.Y.Z or uninstalled` as the two options).  Is the resolver really doing what it says there?  That is, is it really only considering the tagged version or uninstalled, and not realizing that the dev branch (or in the case of a package, `vX.Y.Z+`) is a possibility?

**Comment by KristofferC** [comment]
Right now, standard libraries only have the current julia build as their "version". That shouldn't matter though because no package has any version requirement on them.

Since you started from scratch, the only thing I can think differ is with your julia install, because we look at it to determine the standard libraries. So... anything special there?

Also, this is strange:

```
 BinDeps [9e28174c] log:
...
 ├─restricted by compatibility requirements with (unknown) [8f399da3] to versions: uninstalled
```

It seems that it is saying that a compatability requirement with LibDl means that BinDeps must be uninstalled...

**Comment by mbauman** [comment]
I don't recall doing anything much beyond `git clone` and `make` there.  You're welcome to check it out: `arctic1:/home/mbauman/julia/julia`.

**Comment by KristofferC** [comment]
Pff, I can add it fine on my own arctic julia but it fails on yours. Wat.

**Comment by StefanKarpinski** [comment]
@mbauman: Did you build with Pkg3 precompilation on or off?

**Comment by KristofferC** [comment]
Off, you can see the warning in the pasted logs.

**Comment by mbauman** [comment]
Off — I've recently been re-building Julia more frequently than I've been using Pkg.

**Comment by StefanKarpinski** [comment]
I wonder if that could make a difference since I assume that @KristofferC, you've got it on. It would be a bug for the behavior to be different, of course, but we do have a new optimizer that could be buggy.

**Comment by KristofferC** [comment]
I also got it off. Could you try it?

**Comment by StefanKarpinski** [comment]
> I also got it off.

In that case it can't be the source of the difference, right?

> Could you try it?

Works fine for me:
```
shell> mkdir NewEnv

shell> cd NewEnv/
/Users/stefan/projects/julia/NewEnv

(v0.7) pkg> init
Initialized project at /Users/stefan/projects/julia/NewEnv/Project.toml

(NewEnv) pkg> add BenchmarkTools
  Updating registry at `~/.julia/registries/Uncurated`
  Updating git-repo `https://github.com/JuliaRegistries/Uncurated.git`
 Resolving package versions...
Downloaded Nullables ────── v0.0.5
Downloaded BenchmarkTools ─ v0.3.0
Downloaded JSON ─────────── v0.17.2
  Updating `Project.toml`
  [6e4b80f9] + BenchmarkTools v0.3.0
  Updating `Manifest.toml`
  [6e4b80f9] + BenchmarkTools v0.3.0
  [34da2185] + Compat v0.65.1
  [682c06a0] + JSON v0.17.2
  [4d1e1d77] + Nullables v0.0.5
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [8bb1440f] + DelimitedFiles
  [8ba89e20] + Distributed
  [b77e0a4c] + InteractiveUtils
  [de555fa4] + IterativeEigensolvers
  [76f85450] + LibGit2
  [8f399da3] + Libdl
  [37e2e46d] + LinearAlgebra
  [d6f4376e] + Markdown
  [a63ad114] + Mmap
  [44cfe95a] + Pkg
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [9e88b42a] + Serialization
  [1a1011a3] + SharedArrays
  [6462fe0b] + Sockets
  [2f01184e] + SparseArrays
  [4607b0f0] + SuiteSparse
  [8dfed614] + Test
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
```

**Comment by mbauman** [comment]
Ok, well, I can no longer reproduce this minimal example.  Since two weeks ago I've rebuilt Julia master and the registry has changed — it's the latter that is important.  Rolling back the Uncurated registry state to what it was back then (e6e720ea35c5b2c69509da6950af1162ae172af8) makes my resolver fail in the same manner again.

**Comment by mbauman** [comment]
A quick bisect on the Uncurated registry state points to [`fc4cb27b1`](https://github.com/JuliaRegistries/Uncurated/commit/fc4cb27b1aaa44f75a1a4c726b6c135af30f27c2) as the commit that changed things.  Good news: it actually changed something pertaining to BenchmarkTools.  Bad news: it didn't change _any_ dependency information — it just registered a new patch.

I hope that's helpful.

**Comment by carlobaldassi** [comment]
I can't reproduce this either, with latest Julia (alpha.47) and the Uncurated registry rolled back to https://github.com/JuliaRegistries/Uncurated/commit/e6e720ea35c5b2c69509da6950af1162ae172af8.
This is what I get:

```
(HelloW) pkg> add BenchmarkTools
  Updating registry at `~/.julia/registries/Uncurated`
┌ Warning: Some registries failed to update:
│     — `~/.julia/registries/Uncurated` — registry detached
└ @ Pkg.API API.jl:116
 Resolving package versions...
Downloaded Nullables ────── v0.0.5
Downloaded JSON ─────────── v0.17.2
Downloaded BenchmarkTools ─ v0.3.1
  Updating `Project.toml`
  [6e4b80f9] + BenchmarkTools v0.3.1
  Updating `Manifest.toml`
  [6e4b80f9] + BenchmarkTools v0.3.1
  [34da2185] + Compat v0.68.0
  [682c06a0] + JSON v0.17.2
  [4d1e1d77] + Nullables v0.0.5
  [2a0f44e3] + Base64 
  [ade2ca70] + Dates 
  [8bb1440f] + DelimitedFiles 
  [8ba89e20] + Distributed 
  [b77e0a4c] + InteractiveUtils 
  [de555fa4] + IterativeEigensolvers 
  [76f85450] + LibGit2 
  [8f399da3] + Libdl 
  [37e2e46d] + LinearAlgebra 
  [d6f4376e] + Markdown 
  [a63ad114] + Mmap 
  [44cfe95a] + Pkg 
  [de0858da] + Printf 
  [3fa0cd96] + REPL 
  [9a3f8284] + Random 
  [9e88b42a] + Serialization 
  [1a1011a3] + SharedArrays 
  [6462fe0b] + Sockets 
  [2f01184e] + SparseArrays 
  [4607b0f0] + SuiteSparse 
  [8dfed614] + Test 
  [cf7118a7] + UUIDs 
  [4ec0a83e] + Unicode 
```

As for #354, I'd welcome any suggestion on how to reproduce this.

About this:

> I suspect that we're hitting a fundamental problem with the current resolver.

In this case, this is not it. The current resolver should never declare spurious incompatibilities, so this kind of error should never really happen. In principle, the worst that should happen is that either it runs for a very long time before it can find a solution, or that it finds a sub-optimal solution if many are possible. I have never actually seen it in my tests, although it's obviously possible. In any case when it mistakenly declares an incompatibility error it's a bug somewhere, either in the engine or in the package information that it gets. In this case I would bet on the latter, in particular because of this:

> Forgive my ignorance here, but is this an issue:
>
> │   ├─possible versions are: 0.7.0 or uninstalled
>│   └─fixed by the MaxSum heuristic to version 0.7.0-DEV.5220
>
> I've seen this in other resolver failures (although typically with a package on a master branch and not > julia itself, listing just vX.Y.Z or uninstalled as the two options). Is the resolver really doing what it says > there? That is, is it really only considering the tagged version or uninstalled, and not realizing that the > dev branch (or in the case of a package, vX.Y.Z+) is a possibility?

This should never happen too, Julia should also always be a fixed package, the "uninstalled" state should never be a possibility, and surely the MaxSum solver should not be involved in its resolution, it should get simplified-out much earlier in the process.

**Comment by mbauman** [comment]
I've confirmed that this is the same problem as #303 and JuliaLang/julia#27482 by doing the `any!` reduction over a `Vector{Bool}` instead — and it works.
